Scripts: Include YAML files in prettification

ockham opened the pull request:
## Description
`prettier` has been able to[ format YAML files since version 1.14](https://prettier.io/blog/2018/07/29/1.14.0.html) (released in July of 2018). We're starting to size up on YAML files (mostly GitHub Actions workflows), so it'd be nice to have them auto-formatted. This PR adds YAML files to the list of extensions formatted by `prettier`. Existing YAML files have already been updated to the new format by #30409.

## How has this been tested?
Verify that GH Actions still pass.

## Types of changes
Developer experience

cc/ @jsnajdr @scinos @tyxla @anomiex @jeherve In case y'all wanna do something similar for Calypso and Jetpack

swissspidy commented:
Interesting, so it also forces 4 spaces for indentation?

gziolo commented on packages/scripts/scripts/format-js.js:
```diff
@@ -104,7 +104,7 @@ if ( fileArgs.length === 0 ) {
 // Converts `foo/bar` directory to `foo/bar/**/*.js`
 const globArgs = dirGlob( fileArgs, {
-	extensions: [ 'js', 'jsx', 'ts', 'tsx' ],
+	extensions: [ 'js', 'jsx', 'ts', 'tsx', 'yml', 'yaml' ],
```
It's `format-js` so I don't think it should be concerned about YAML. Can we create another script that would cover config formats like YAML or JSON?

ockham replied:
Yeah, I was worried that that could be of concern 😅 

> Can we create another script that would cover config formats like YAML or JSON?

IDK, where would we draw the line? We're already using the same script for `.ts(x)`; I think at least JSON would also be "close enough" to JS to be included.

I get that YAML is quite different, but the drawback is that we'd end up with two separate scripts that do pretty much the exact same thing -- running `prettier` on a bunch of files. Which can be done -- potentially by extracting some logic into a shared module -- but it's going to be more than a one-line change as here 😬 

Would we consider renaming `format-js` instead? 🤔

gziolo replied:
I don't think we should consider the number of lines to update as a deciding factor. We need to keep in mind backward compatibility and flexibility of usage. At the moment, it'd be a breaking change for those who use `format-js` because it suddenly would format more files than before. What if someone wants to opt-out of it?

youknowriad replied:
what if we rename the command just `format` ?

scinos replied:
FYI, in Calypso we moved to `eslint` as the "format engine", used to format all types of files: `.js`, `.ts`, `.md`... It uses `prettier` as a plugin (so you get the same formatting you'd get with just `prettier`), plus it allows us to use all `eslint` goodies like custom rules, disable rules per line/file, custom reporters...

gziolo replied:
We can keep an alias for backward compatibility and rename, but still, the question remains, what if someone wants to format only JS or only YAML? If we sort it out, then everything is fine.

youknowriad replied:
>  what if someone wants to format only JS or only YAML? If we sort it out, then everything is fine.

Why would someone want that? I mean it's definitely possible but 95% of the cases, people just want to have guidelines for all their code base and `wp-scripts` is about a good opinionated setup so I'm personally in favor of removing all configs and possibilities from folks and just simplifying everything. `format` command seems to go in that direction for me.

youknowriad replied:
An alias is a good idea to avoid distruption but we can also make a major version change. If I had to choose I'd do both  and deprecate the old.

ockham replied:
Agree with Riad. IMO, this approach also better reflects the idea of `prettier`: It takes care of formatting for you so you don't have to worry about _how_ exactly you want to fine-tune whitespace etc -- going so far as to even include new file formats, once it supports them. (Meaning that if you would run `prettier` directly, I think it'd also just start formatting those new files unless you explicitly excluded them.)

We could make it a breaking change: Bump `wp/scripts` to a new major version, and rename `format-js` to `format`.

gziolo replied:
You can give it a try to add `format` that covers all files that Prettier supports. I hope it is compatible with `stylelint` for SCSS files.

For `format-js`, it would be mostly about the convenience of developers who already have it integrated into the project. `npx wp-create-block` uses `wp-scripts format-js` so it might impact a larger group. We can remove it from the documentation and replace it with `format`. It would be simply nice to keep an alias with a deprecation warning.

ockham replied:
> It would be simply nice to keep an alias with a deprecation warning.

Sounds good to me!

ockham replied:
I've changed the script name and added a deprecation warning per the latest commits.

gziolo commented:
In general, I love the idea. I'd like to see all file formats covered one day. At least for Gutenberg.

> Interesting, so it also forces 4 spaces for indentation?

It would be simpler to override some settings if JS wouldn't be concerned in the mix.

As noted in the comment, we probably should create its own group that is concerned with files used as configs like YAML or JSON.

Any thoughts on how we can ensure consistent formatting with CI?

ockham commented:
> Interesting, so it also forces 4 spaces for indentation?

Yeah, I think it basically just applies the same `prettier-config` that we're using for JS already: https://github.com/WordPress/gutenberg/blob/2cdc0f67eccf35d227b2b7de48249b4d0c2fd6ab/packages/prettier-config/lib/index.js#L14-L15

Personally, I'm fine with whatever, as long as it's auto-formatted 😬  and I like to keep a minimum of config files.
Would y'all prefer different indentation?

gziolo commented:
> Personally, I'm fine with whatever, as long as it's auto-formatted 😬 and I like to keep a minimum of config files.
Would y'all prefer different indentation?

Agreed, it's simpler to keep one set of rules for everything. In case we want some tweaks, you can override settings with CLI flags.

gziolo commented on packages/create-block/CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
+### Breaking Changes
+
+-   Rename `format:js` script to `format` ([#30240](https://github.com/WordPress/gutenberg/pull/30240)).
+-   Include YAML files when formatting files with `format` ([#30240](https://github.com/WordPress/gutenberg/pull/30240)).
```
You can skip YAML part here, it's relevant only for `@wordpress/scripts`.

ockham replied:
01e75fd8f6

ockham commented:
This should be ready for another look: I've now renamed `format-js` to `format`, and added a thin `format-js` wrapper that includes a deprecation note.

gziolo commented on package.json:
```diff
@@ -292,7 +293,7 @@
 			"wp-scripts lint-style"
 		],
 		"*.{js,ts,tsx}": [
```
Should it cover YAML now?
```suggestion
		"*.{js,ts,tsx,yml}": [
```

ockham replied:
Ah, missed that one! Thanks!

gziolo reviewed (commented):
It would be great to include other formats in the same npm release so we don't have to introduce them every 2 weeks and break existing workflows. Otherwise, I think it's close to ready. I haven't had a chance to test it yet.

ockham commented:
> It would be great to include other formats in the same npm release so we don't have to introduce them every 2 weeks and break existing workflows. Otherwise, I think it's close to ready. I haven't had a chance to test it yet.

I tried adding JSON, but I'd rather do it in a separate PR for the following reasons:
- This PR already touches 30 files.
- If we want to enable JSON support for `lint-staged`, more work is needed, as that also runs the `lint-js` script, which will currently flag JSON files as invalid JS -- it'll likely require some more tweaking of the ESLint options or config. Arguably, this _should_ be done in the same step as enabling `prettier` for JSON files.
- Even if we don't enable JSON support for `lint-staged`, running `prettier` on our JSON files for the first time touches _a lot_ of files. Among them are the snapshots in `packages/e2e-tests/fixtures/blocks/`, which probably deserve some extra attention.

In conclusion, I'm worried more about scope creep for this now still fairly-contained PR more than about adding support for JSON in a separate step (especially after we've renamed `format-js` to `format`) 😬  

Are there any other file formats you had in mind? For reference, I've found a list of supported formats [here](https://prettier.io/docs/en/index.html).

<details>
<summary>Diff for enabling JSON</summary>

```diff
diff --git a/package.json b/package.json
index 99d31594a5..c8aeb17891 100644
--- a/package.json
+++ b/package.json
@@ -292,7 +292,7 @@
                "*.scss": [
                        "wp-scripts lint-style"
                ],
-               "*.{js,ts,tsx,yml}": [
+               "*.{js,json,ts,tsx,yml}": [
                        "wp-scripts format",
                        "wp-scripts lint-js"
                ],
diff --git a/packages/scripts/CHANGELOG.md b/packages/scripts/CHANGELOG.md
index b31a5c8ab2..0fb7a81378 100644
--- a/packages/scripts/CHANGELOG.md
+++ b/packages/scripts/CHANGELOG.md
@@ -5,7 +5,7 @@
 ### Breaking Changes
 
 -   Rename `format-js` script to `format` ([#30240](https://github.com/WordPress/gutenberg/pull/30240)).
--   Include YAML files when formatting files with `format` ([#30240](https://github.com/WordPress/gutenberg/pull/30240)).
+-   Include JSON and YAML files when formatting files with `format` ([#30240](https://github.com/WordPress/gutenberg/pull/30240)).
 -   The bundled `css-loader` dependency has been updated from requiring `^3.5.2` to requiring `^5.1.3` ([#27821](https://github.com/WordPress/gutenberg/pull/27821)).
 -   The bundled `mini-css-extract-plugin` dependency has been updated from requiring `^0.9.0` to requiring `^1.3.9` ([#27821](https://github.com/WordPress/gutenberg/pull/27821)).
 -   The bundled `postcss-loader` dependency has been updated from requiring `^3.0.0` to requiring `^4.2.0` ([#27821](https://github.com/WordPress/gutenberg/pull/27821)).
diff --git a/packages/scripts/scripts/format.js b/packages/scripts/scripts/format.js
index 6dd655bdab..40547216e1 100644
--- a/packages/scripts/scripts/format.js
+++ b/packages/scripts/scripts/format.js
@@ -104,7 +104,7 @@ if ( fileArgs.length === 0 ) {
 
 // Converts `foo/bar` directory to `foo/bar/**/*.js`
 const globArgs = dirGlob( fileArgs, {
-       extensions: [ 'js', 'jsx', 'ts', 'tsx', 'yml', 'yaml' ],
+       extensions: [ 'js', 'json', 'jsx', 'ts', 'tsx', 'yml', 'yaml' ],
 } );
 
 const result = spawn(
```
</details>

gziolo commented:
> I tried adding JSON, but I'd rather do it in a separate PR for the following reasons:
> In conclusion, I'm worried more about scope creep for this now still fairly-contained PR more than about adding support for JSON in a separate step (especially after we've renamed format-js to format) 

There are ways to mitigate it. You can always apply formatting in separate PRs, one for each file extension type before we introduce more strict integration. I guess that's the only concern in terms of the size of this PR.

> Are there any other file formats you had in mind? For reference, I've found a list of supported formats here.

We also use JSON and SCSS from the list of supported types of files.

ockham commented:
> > I tried adding JSON, but I'd rather do it in a separate PR for the following reasons:
> > In conclusion, I'm worried more about scope creep for this now still fairly-contained PR more than about adding support for JSON in a separate step (especially after we've renamed format-js to format)
> 
> There are ways to mitigate it. You can always apply formatting in separate PRs, one for each file extension type before we introduce more strict integration. I guess that's the only concern in terms of the size of this PR.

Well, there goes my plan for a quick PR to get YAML formatted, and iterate later on it to include other formats in order to keep the cognitive load smaller 😬 
FWIW, I've filed #30409 for now for just the YAML formatting changes.

> > Are there any other file formats you had in mind? For reference, I've found a list of supported formats here.
> 
> We also use JSON and SCSS from the list of supported types of files.

If we really want to inflate the scope of this to include different formats that currently have their own linters etc, we might want to consider following a similar strategy as @scinos [mentioned](https://github.com/WordPress/gutenberg/pull/30240#discussion_r601482283): 

> FYI, in Calypso we moved to `eslint` as the "format engine", used to format all types of files: `.js`, `.ts`, `.md`... It uses `prettier` as a plugin (so you get the same formatting you'd get with just `prettier`), plus it allows us to use all `eslint` goodies like custom rules, disable rules per line/file, custom reporters...

ockham commented:
> If we really want to inflate the scope of this to include different formats that currently have their own linters etc, we might want to consider following a similar strategy as @scinos [mentioned](https://github.com/WordPress/gutenberg/pull/30240#discussion_r601482283):
> 
> > FYI, in Calypso we moved to `eslint` as the "format engine", used to format all types of files: `.js`, `.ts`, `.md`... It uses `prettier` as a plugin (so you get the same formatting you'd get with just `prettier`), plus it allows us to use all `eslint` goodies like custom rules, disable rules per line/file, custom reporters...

This seems to be the relevant Calypso PR: https://github.com/Automattic/wp-calypso/pull/40118

ockham commented:
> > If we really want to inflate the scope of this to include different formats that currently have their own linters etc, we might want to consider following a similar strategy as @scinos [mentioned](https://github.com/WordPress/gutenberg/pull/30240#discussion_r601482283):
> > > FYI, in Calypso we moved to `eslint` as the "format engine", used to format all types of files: `.js`, `.ts`, `.md`... It uses `prettier` as a plugin (so you get the same formatting you'd get with just `prettier`), plus it allows us to use all `eslint` goodies like custom rules, disable rules per line/file, custom reporters...
> 
> This seems to be the relevant Calypso PR: [Automattic/wp-calypso#40118](https://github.com/Automattic/wp-calypso/pull/40118)

Hmm, `prettier` kinda seems to recommend against using it as an `eslint` plugin: https://prettier.io/docs/en/integrating-with-linters.html

gziolo commented:
> Hmm, prettier kinda seems to recommend against using it as an eslint plugin: https://prettier.io/docs/en/integrating-with-linters.html

@ntwb tried that approach in the past and it didn't work very well. Related PR and the conclusions can be found in https://github.com/WordPress/gutenberg/pull/4628#issuecomment-359362473 and https://github.com/WordPress/gutenberg/pull/4628#issuecomment-360071509.

I can't find the exact comment, maybe it's another PR but there were some issues with `eslint --fix` that produced different output on subsequent runs.

gziolo commented:
One more thing, we will need to add missing handling for `.prettierignore`. At the moment I see we use a fallback for `.eslintignore` which is perfectly fine for JS files, but won't scale when new formats are added:

https://github.com/WordPress/gutenberg/blob/719d575bea0dea1f2215f239cf4bece6aeb714fd/packages/scripts/scripts/format-js.js#L83-L91

We can copy `.eslintignore` for start:
https://github.com/WordPress/gutenberg/blob/719d575bea0dea1f2215f239cf4bece6aeb714fd/packages/scripts/config/.eslintignore#L1-L3

`node_modules` is ignored by default, so most likely it can be omitted.

gziolo commented on packages/create-block/CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+### Breaking Changes
```
I was thinking about that, and it seems to me like it should be classified as an `Enhancement`. My reasoning is that it doesn't impact existing projects, but instead changes how the newly scaffolded project is configured.

gziolo commented:
I opened PRs that try to format:
- markdown files https://github.com/WordPress/gutenberg/pull/30715
- JSON files https://github.com/WordPress/gutenberg/pull/30714

ockham committed:
Scripts: Include YAML files in prettification

ockham committed:
Move create-block's Changelog entry to Enhancement

gziolo commented on packages/scripts/README.md:
```diff
@@ -128,7 +128,7 @@ _Flags_:
-### `format-js`
+### `format`
 
 It helps to enforce coding style guidelines for your JavaScript files by formatting source code in a consistent way.
```
```suggestion
It helps to enforce coding style guidelines for your files (JavaScript, YAML) by formatting source code in a consistent way.
```

gziolo commented:
Two more comments to address and we can land it:
- better way to list ignored files https://github.com/WordPress/gutenberg/pull/30240#issuecomment-817515651
- tiny documentation clarification https://github.com/WordPress/gutenberg/pull/30240#discussion_r613103517

If you prefer to tackle ignored files separately, then we can split it, too.

ockham committed:
Small doc tweak

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

ockham commented:
Thanks for approving, @gziolo! ❤️ 

> Two more comments to address and we can land it:
> 
> * better way to list ignored files [#30240 (comment)](https://github.com/WordPress/gutenberg/pull/30240#issuecomment-817515651)

I'll file a separate PR for this. I tried it locally, and it surprisingly changes the formatting of some files in the `e2e-tests` package. I'd like to look into this separately in order to unblock this PR 🙂 

> * tiny documentation clarification [#30240 (comment)](https://github.com/WordPress/gutenberg/pull/30240#discussion_r613103517)

✅ 

I will include some more indentation changes to `.yml` files with this PR -- seems like the `node-version:` lines got the wrong indentation by #29078.

ockham committed:
Fix indendation in YAML files
